libslide/src/bignum/sub.rs

```diff
@@ -0,0 +1,148 @@
+#![allow(clippy::suspicious_arithmetic_impl)]
+use crate::bignum::utils::is_smaller;
+use crate::bignum::Bignum;
+use std::convert::TryInto;
+use std::mem;
+use std::ops;
+
+impl ops::Sub for Bignum {
+    type Output = Bignum;
+
+    fn sub(self, other: Bignum) -> Bignum {
+        // make the decimal vectors the same size
+        // fill lhs and rhs vectors as well
+        let mut lhs_dec: Vec<u8>;
+        let mut res_dec: Vec<u8>;
+        let self_size: usize = self.dec.len();
+        let other_size: usize = other.dec.len();
+        let (mut res_int, mut lhs_int) = if self_size > other_size {
+            res_dec = self.dec;
+            lhs_dec = vec![0; self_size];
+            lhs_dec[..other_size].clone_from_slice(&other.dec[..other_size]);
+            (self.int, other.int)
+        } else {
+            res_dec = other.dec;
+            lhs_dec = vec![0; other_size];
+            lhs_dec[..self_size].clone_from_slice(&self.dec[..self_size]);
+            (other.int, self.int)
+        };
+
+        // now we need to make the bigger number the res dec
+        if is_smaller(&lhs_dec, &res_dec, &lhs_int, &res_int) {
+            mem::swap(&mut lhs_dec, &mut res_dec);
+            mem::swap(&mut lhs_int, &mut res_int);
+        }
+
+        // 1. Decimal
+
+        let mut res: i8;
+        let mut carry: i8 = 0;
+        for i in (0..res_dec.len()).rev() {
+            res = res_dec[i] as i8 - lhs_dec[i] as i8 - carry;
+            // not sure if you can use a let if here because of shadowing
+            // but maybe you know
+            if res < 0 {
+                res += 10;
+                carry = 1;
+            } else {
+                carry = 0;
+            }
+            res_dec[i] = res.try_into().unwrap();
+        }
+
+        // 2. Integer
+
+        for i in 0..lhs_int.len() {
+            res = res_int[i] as i8 - lhs_int[i] as i8 - carry;
+            if res < 0 {
+                res += 10;
+                carry = 1;
+            } else {
+                carry = 0;
+            }
+            res_int[i] = res.try_into().unwrap();
```

Comment:
Yeah let's make res_int a `Vec<i8>`

Reply:
So technically this wont change the casting. Because either way res_int has to be a u8 in Bignum. so we would have to convert to build it and then convert back to store it

Reply:
that's fine, it's cleaner than doing it each time in a loop imo

Reply:
so how do you wanna build res_dec. fill with zeros and then cast? I think its equally ugly

Reply:
also converting to a vector creates a  new allocation of a vector. I think I like this way better

Reply:
just do it this way: https://stackoverflow.com/questions/59707349/cast-vector-of-i8-to-vector-of-u8-in-rust